event_model/__init__.py

```diff
@@ -180,6 +180,103 @@ def bulk_datum(self, doc):
         self.datum_page(bulk_datum_to_datum_page(doc))
 
 
+class SingleRunDocumentRouter(DocumentRouter):
+    """
+    A DocumentRouter intended to process events from exactly
+    one run.
+    """
+    def __init__(self):
+        self._start_doc = None
+        self._descriptors = dict()
+
+    def __call__(self, name, doc, validate=False):
+        """
+        Keep track of the start document and descriptor documents
+        passed to this SingleRunDocumentRouter.
+
+        Send documents to the superclass for processing.
+
+        Parameters
+        ----------
+        name : string
+        doc : dict
+        validate : boolean
+            Apply jsonschema validation to the documents coming *out*. This is
+            False by default.
+
+        Returns
+        -------
+        name, output_doc : string, dict
+            The same name as what was passed in, and a doc that may be the same
+            instance as doc, a copy of doc, or a different dict altogether.
+        """
+        if name == 'start':
+            if self._start_doc is None:
+                self._start_doc = doc
+            else:
+                raise EventModelError(
+                    'SingleRunDocumentRouter received a second start document: uid {doc["uid"]}'
+                )
+        elif name == 'descriptor':
+            if doc['run_start'] == self._start_doc['uid']:
+                self._descriptors[doc['uid']] = doc
+            else:
+                raise EventModelError(
+                    'SingleRunDocumentRouter received a descriptor associated with a different start document'
+                )
+
+        return super().__call__(name, doc, validate)
+
+    def get_start(self):
+        """Convenience method returning the start document for the associated run.
+
+        If no start document has been processed EventModelError will be raised.
+
+        Returns
+        -------
+        start document : dict
+        """
+        if self._start_doc is None:
+            raise EventModelError('SingleRunDocumentRouter has not processed a start document yet')
+
+        return self._start_doc
+
+    def get_descriptor(self, doc):
+        """Convenience method returning the descriptor associated with the specified document.
+
+        Parameters
+        ----------
+        doc : dict
+            event-model document
+
+        Returns
+        -------
+        descriptor document : dict
+        """
+        if 'descriptor' not in doc:
+            raise ValueError(f'document is not associated with a descriptor:\n{doc}')
```

Comment:
All errors directly raised from inside `event_model` should be subclasses of `EventModelError`. This one could be `EventModelValueError`, which I believe is already defined below.

This someone picky pattern is one that we apply in event-model and the suitcase libraries, but have not taken the trouble to retroactively apply to the older libraries. The purpose is to make it clear whether event-model did a thing that cause _another_ library to raise (including a standard Python library) or whether the problem is directly inside event-model

Reply:
I'll be honest, I just used a couple different exceptions to see which ones would stick.